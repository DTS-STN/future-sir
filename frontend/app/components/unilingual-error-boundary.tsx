import { Links, Meta, Scripts } from 'react-router';

import { Trans, useTranslation } from 'react-i18next';

import type { Route } from '../+types/root';

import { AppLink } from '~/components/links';
import { PageTitle } from '~/components/page-title';
import { isAppError } from '~/errors/app-error';
import { useLanguage } from '~/hooks/use-language';

/**
 * A unilingual error boundary that renders appropriate error messages in the current language.
 *
 * **Important Note:**
 *
 * React Router error boundaries should be designed to be as robust as possible.
 * If an error boundary itself throws an error, there's no subsequent error
 * boundary to catch and render it, potentially leading to infinite error loops.
 */
export function UnilingualErrorBoundary({ actionData, error, loaderData, params }: Route.ErrorBoundaryProps) {
  const { currentLanguage } = useLanguage();
  const { t } = useTranslation(['gcweb']);

  return (
    <html lang={currentLanguage}>
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <header className="border-b-[3px] border-slate-700 print:hidden">
          <div id="wb-bnr">
            <div className="container flex items-center justify-between gap-6 py-2.5 sm:py-3.5">
              <AppLink to="https://canada.ca/">
                <img
                  className="h-8 w-auto"
                  src={`https://www.canada.ca/etc/designs/canada/wet-boew/assets/sig-blk-${currentLanguage}.svg`}
                  alt={t('gcweb:header.govt-of-canada.text')}
                  width="300"
                  height="28"
                  decoding="async"
                />
              </AppLink>
            </div>
          </div>
        </header>
        <main className="container">
          <PageTitle className="my-8">
            <span>{t('gcweb:server-error.page-title')}</span>
            <small className="block text-2xl font-normal text-neutral-500">
              {t('gcweb:server-error.page-subtitle', { statusCode: isAppError(error) ? error.httpStatusCode : 500 })}
            </small>
          </PageTitle>
          <p className="mb-8 text-lg text-gray-500">{t('gcweb:server-error.page-message')}</p>
          {isAppError(error) && (
            <ul className="list-disc pl-10 text-gray-800">
              <li>
                <Trans
                  i18nKey="gcweb:server-error.error-code"
                  components={{ span: <span className="font-mono" />, strong: <strong className="font-semibold" /> }}
                  values={{ errorCode: error.errorCode }}
                />
              </li>
              <li>
                <Trans
                  i18nKey="gcweb:server-error.correlation-id"
                  components={{ span: <span className="font-mono" />, strong: <strong className="font-semibold" /> }}
                  values={{ correlationId: error.correlationId }}
                />
              </li>
            </ul>
          )}
        </main>
        <footer id="wb-info" tabIndex={-1} className="mt-8 bg-stone-50 print:hidden">
          <div className="container flex items-center justify-end gap-6 py-2.5 sm:py-3.5">
            <div>
              <h2 className="sr-only">{t('gcweb:footer.about-site')}</h2>
              <div>
                <img
                  src="https://www.canada.ca/etc/designs/canada/wet-boew/assets/wmms-blk.svg"
                  alt={t('gcweb:footer.gc-symbol')}
                  width={300}
                  height={71}
                  className="h-10 w-auto"
                />
              </div>
            </div>
          </div>
        </footer>
        <Scripts nonce={loaderData?.nonce} />
        <script //
          nonce={loaderData?.nonce}
          src={`/api/client-env?v=${loaderData?.clientEnvRevision}`}
          suppressHydrationWarning={true}
        />
      </body>
    </html>
  );
}
