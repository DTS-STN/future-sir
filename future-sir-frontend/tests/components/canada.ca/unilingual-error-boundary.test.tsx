import { createRoutesStub } from 'react-router';

import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { UnilingualErrorBoundary } from '~/components/canada.ca/unilingual-error-boundary';

vi.mock('react-i18next');

describe('UnilingualErrorBoundary', () => {
  it('should render the unilingual error boundary', () => {
    vi.spyOn(console, 'error').mockImplementation(() => {});

    const RoutesStub = createRoutesStub([
      {
        path: '/en/public',
        Component: () => <UnilingualErrorBoundary params={{}} error={new Error('Something went wrong')} />,
      },
    ]);

    render(<RoutesStub initialEntries={['/en/public']} />);
    expect(document.documentElement.outerHTML).toMatchSnapshot('expected html');
  });
});
