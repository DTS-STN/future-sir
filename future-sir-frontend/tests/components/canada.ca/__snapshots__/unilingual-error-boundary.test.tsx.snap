// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnilingualErrorBoundary > should render the unilingual error boundary > expected html 1`] = `"<html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1"></head><body vocab="http://schema.org/" typeof="WebPage"><header class="border-b-[3px] border-slate-700 print:hidden"><div id="wb-bnr"><div class="container flex items-center justify-between gap-6 py-2.5 sm:py-3.5"><div property="publisher" typeof="GovernmentOrganization"><a property="url" href="https://canada.ca/"><img class="h-8 w-auto" src="https://www.canada.ca/etc/designs/canada/wet-boew/assets/sig-blk-en.svg" alt="gcweb:header.govt-of-canada.text" property="logo" width="300" height="28" decoding="async"></a><meta property="name" content="gcweb:header.govt-of-canada.text"><meta property="areaServed" typeof="Country" content="Canada"><link property="logo" href="https://www.canada.ca/etc/designs/canada/wet-boew/assets/wmms-blk.svg"></div></div></div></header><main class="container" property="mainContentOfPage" resource="#wb-main" typeof="WebPageElement"><h1 id="wb-cont" tabindex="-1" class="font-lato text-3xl font-bold focus-visible:ring my-8" property="name"><span>gcweb:server-error.page-title</span><small class="block text-2xl font-normal text-neutral-500">gcweb:server-error.page-subtitle</small></h1><p class="mb-8 text-lg text-gray-500">gcweb:server-error.page-message</p></main><footer id="wb-info" tabindex="-1" class="mt-8 bg-stone-50 print:hidden"><div class="container flex items-center justify-end gap-6 py-2.5 sm:py-3.5"><div><h2 class="sr-only">gcweb:footer.about-site</h2><div><img src="https://www.canada.ca/etc/designs/canada/wet-boew/assets/wmms-blk.svg" alt="gcweb:footer.gc-symbol" width="300" height="71" class="h-10 w-auto"></div></div></div></footer><link rel="modulepreload" href=""><link rel="modulepreload" href="build/stub-path-to-module.js"><script> </script><script type="module" async=""> </script><script>window.__appEnvironment = undefined</script></body></html>"`;
